Ensure CI is run in FIPS 140 approved only mode

jkakavas opened the pull request:
We were depending on the BouncyCastle FIPS own mechanics to set
itself in approved only mode since we run with the Security
Manager enabled. The check during startup seems to happen before we
set our restrictive SecurityManager though in
org.elasticsearch.bootstrap.Elasticsearch , and this means that
BCFIPS would not be in approved only mode, unless explicitly
configured so.

This commit sets the appropriate JVM property to explicitly set
BCFIPS in approved only mode in CI and adds tests to ensure that we
will be running with BCFIPS in approved only mode when we expect to. 
It also sets `xpack.security.fips_mode.enabled` to true for all test clusters
used in fips mode and sets the distribution to the default one. It adds a 
password to the elasticsearch keystore for all test clusters that run in fips
mode.
Moreover, it changes a few unit tests where we would use bcrypt even in
FIPS 140 mode. These would still pass since we are bundling our own
bcrypt implementation, but are now changed to use FIPS 140 approved
algorithms instead for better coverage.

It also addresses a number of tests that would fail in approved only mode 
Mainly:

- Tests that use PBKDF2 with a password less than 112 bits (14char). We 
elected to change the passwords used everywhere to be at least 14 
characters long instead of mandating
the use of `pbkdf2_stretch` because both `pbkdf2` and
`pbkdf2_stretch` are supported and allowed in fips mode and it makes sense
to test with both. We could possibly figure out the password algorithm used
for each test and adjust password length accordingly only for `pbkdf2` but 
there is little value in that. It's good practice to use strong passwords so if
our docs and tests use longer passwords, then it's for the best. The approach
is brittle as there is no guarantee that the next test that will be added won't
use a short password, so we add some testing documentation too. 
This leaves us with a possible coverage gap since we do support passwords
as short as 6 characters but we only test with > 14 chars but the 
validation itself was not tested even before. Tests can be added in a followup,
outside of fips related context. 

- Tests that use a PKCS12 keystore and were not already muted.

- Tests that depend on running test clusters with a basic license or 
using the OSS distribution as FIPS 140 support is not available in 
neither of these.

Finally, it adds some information around FIPS 140 testing in our testing
documentation reference so that developers can hopefully keep in 
mind fips 140 related intricacies when writing/changing docs.

jkakavas committed:
Ensure CI is run in FIPS 140 approved only mode

We were depending on the BouncyCastle FIPS own mechanics to set
itself in approved only mode since we run with the Security
Manager enabled. The check during startup seems to happen before we
set our restrictive SecurityManager though in
org.elasticsearch.bootstrap.Elasticsearch , and this means that
BCFIPS would not be in approved onlu mode, unless explicitly
configured so.

This commit sets the appropriate JVM property to explicitly set
BCFIPS in approved only mode and addresses a number of tests that
would fail in approved mode. Mainly:
- Tests that use PBKDF2 with a password less than 112 bits (14char)
- Tests that use a PKCS12 keystore and were not already muted.

It also changes a few unit tests where we would use bcrypt even in
FIPS 140 mode. These would still pass since we are bundling our own
bcrypt implementation, but are now changed to use FIPS 140 approved
algorithms instead for better coverage.

elasticmachine commented:
Pinging @elastic/es-security (:Security/Security)

ywangd commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/Fips140ProviderVerificationTests.java:
```diff
@@ -0,0 +1,32 @@
+
+public class Fips140ProviderVerificationTests extends ESTestCase {
+
+    public void testBcFipsProviderInUse(){
+        if (inFipsJvm()){
+            assertThat(Security.getProviders().length > 0, equalTo(true));
+            assertThat(Security.getProviders()[0].getName(), containsString("BCFIPS"));
+        }
+    }
+
+    public void testInApprovedOnlyMode(){
+        if (inFipsJvm()){
+            assertThat(CryptoServicesRegistrar.isInApprovedOnlyMode(), equalTo(true));
+        }
+    }
```
I think these tests can have better coverage if we move them into `ESTestCase#inFipsJvm` so that they are always checked if fips is enabled.

tvernum replied:
I think it's better here. I feel like this is more explicit in its own test that has its own distinct purpose rather than burying it in another test (case).

We might get more coverage by moving them, but I don't understand the value of that - we only need to run them once to get the value we need.

ywangd replied:
I think it has some value, maybe not for CI, but for individual tests when you need run them locally so that you can be sure FIPS is always correctly configured by the property passed from CLI.

jkakavas replied:
> I think it has some value, maybe not for CI, 

The purpose is to cover CI ,do that we don't make any change in the feature that would lead to us not running CI in non approved mode. 

I think the clarity of it being in it's own class makes sense and we should leave it here

ywangd commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/ClusterPrivilegeIntegrationTests.java:
```diff
@@ -79,8 +79,9 @@ protected String configRoles() {
-            randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9")).hash(new SecureString("passwd".toCharArray())));
+        final Hasher passwdHasher = inFipsJvm() ? Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000")) :
+            Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9")) ;
+        final String usersPasswdHashed = new String(passwdHasher.hash(SecuritySettingsSourceField.TEST_PASSWORD_SECURE_STRING));
```
The password was `passwd` which is different from `SecuritySettingsSourceField.TEST_PASSWORD_SECURE_STRING`. This and a few similar changes in other places are causing CI failure for this PR. I think the fix should be updating `AbstractPrivilegeTestCase#setUser` to use the new password as well.

ywangd commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/CreateDocsIndexPrivilegeTests.java:
```diff
@@ -22,14 +22,14 @@
             "    - names: '*'\n" +
             "      privileges: [ all ]\n" +
-        "create_doc_role:\n" +
+            "create_doc_role:\n" +
```
Nit: The indentation is now inconsistent with `all_indices_role`. I do think the change here looks better. But I'd suggest we make the same change to `all_indices_role` as well.

ywangd commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/CreateDocsIndexPrivilegeTests.java:
```diff
@@ -22,14 +22,14 @@
     private static final String USERS_ROLES =
         "all_indices_role:admin\n" +
-        "create_doc_role:" + CREATE_DOC_USER + "\n";
+            "create_doc_role:" + CREATE_DOC_USER + "\n";
```
Nit: indentation inconsistency.

ywangd commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreCacheTests.java:
```diff
@@ -249,12 +249,12 @@ public void testRolesCacheIsClearedWhenPrivilegesIsChanged() {
             .roles(testRole)
-            .password(new SecureString("password".toCharArray()),
-                Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt9", "bcrypt8", "bcrypt")))
+            .password(new SecureString("longerpassword".toCharArray()), hasher)
```
The same password change needs to be made below (L263) when constructing the `Client`.

ywangd reviewed (commented):
I left a few comments. Woud like to have another round later. Thanks!

tvernum commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/Fips140ProviderVerificationTests.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public void testInApprovedOnlyMode(){
+        if (inFipsJvm()){
```
Nit:
```suggestion
        if (inFipsJvm()) {
```

jkakavas replied:
Depending on tools sucks when the tooling fails :/

tvernum commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/Fips140ProviderVerificationTests.java:
```diff
@@ -0,0 +1,32 @@
+
+public class Fips140ProviderVerificationTests extends ESTestCase {
+
+    public void testBcFipsProviderInUse(){
```
Nit
```suggestion
    public void testBcFipsProviderInUse() {
```

tvernum commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/Fips140ProviderVerificationTests.java:
```diff
@@ -0,0 +1,32 @@
+public class Fips140ProviderVerificationTests extends ESTestCase {
+
+    public void testBcFipsProviderInUse(){
+        if (inFipsJvm()){
```
```suggestion
        if (inFipsJvm()) {
```

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/file/FileUserPasswdStoreTests.java:
```diff
@@ -51,6 +52,11 @@
+    @BeforeClass
+    public static void disableInFips(){
+        assumeFalse("We are using hashing algorithms that won't be available in FIPS-140 mode", inFipsJvm());
+    }
```
I don't think we should disable this whole test in FIPS. It seems like an unnecessary reduction in test coverage.

jkakavas replied:
Sure, that was a lazy decision

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlTestCase.java:
```diff
@@ -85,13 +85,21 @@ public static void restoreLocale() {
                 break;
             case "RSA":
-                keySize = randomFrom(1024, 2048, 4096);
+                if (inFipsJvm()){
```
```suggestion
                if (inFipsJvm()) {
```

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlTestCase.java:
```diff
@@ -85,13 +85,21 @@ public static void restoreLocale() {
                 break;
             default:
-                keySize = randomFrom(1024, 2048);
+                keySize = randomFrom(2048);
```
```suggestion
                keySize = 2048;
```

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/support/CachingUsernamePasswordRealmTests.java:
```diff
@@ -65,7 +66,11 @@ public void stop() {
     public void testCacheSettings() {
-        String cachingHashAlgo = randomFrom(Hasher.getAvailableAlgoCacheHash());
+        List<String> availableCacheAlgos = Hasher.getAvailableAlgoCacheHash();
+        if (inFipsJvm()){
```
```suggestion
        if (inFipsJvm()) {
```

jkakavas committed:
fix passwords for tests

jkakavas committed:
Re-enable FileUserPasswdStoreTests

jkakavas committed:
unused import

jkakavas committed:
Adjust a couple of more tests

ywangd commented:
I could be missing something, but is `xpack.security.fips_mode.enabled` set to `true` anywhere when `-Dtests.fips.enabled=true`. It seems to me that we should at least have it enabled for testClusters?

jkakavas committed:
Fix integration tests to use long passwords

jkakavas committed:
Set fips_mode.enabled to true and set a defaut compliant password hashing algo for testclusters

jkakavas committed:
revert test debugging change

jkakavas committed:
Add password to keystores and make sure passwords are longer than 14 chars

jkakavas committed:
Merge remote-tracking branch 'origin/master' into fips-approved-mode-tests

jkakavas committed:
Use long passwords in rest tests

jkakavas committed:
Add keystore password when in fips mode, and fix some passwords

jkakavas committed:
Merge remote-tracking branch 'origin/master' into fips-approved-mode-tests

jkakavas committed:
Skip REST YAML tests in FIPS 140 mode

Currently we don't have a way to selectively mute REST YAML tests
in FIPS 140 mode.

This commit introduces a new feature (fips_140) that can be used
in skip blocks to allow that.

jkakavas committed:
Disable java rest tests for IDP in FIPS 140 mode

jkakavas committed:
Mute tests using PKCS#12

jkakavas committed:
Use BuildParams instead of parsing the sysprop again

BigPandaToo commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/ClusterPrivilegeIntegrationTests.java:
```diff
@@ -79,8 +79,9 @@ protected String configRoles() {
-        final String usersPasswdHashed = new String(Hasher.resolve(
-            randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9")).hash(new SecureString("passwd".toCharArray())));
+        final Hasher passwdHasher = inFipsJvm() ? Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000")) :
+            Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9")) ;
```
Should we add "pbkdf2_stretch"for completeness?

jkakavas replied:
yes, it's in my to do list. When I added the commit there was no pbkf2_stretch

BigPandaToo commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/CreateDocsIndexPrivilegeTests.java:
```diff
@@ -43,8 +43,9 @@ protected String configRoles() {
-        final String usersPasswdHashed = new String(Hasher.resolve(
-            randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9")).hash(new SecureString("passwd".toCharArray())));
+        final Hasher passwdHasher = inFipsJvm() ? Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000")) :
+            Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9"));
```
Same as above. "pbkdf2_stretch"?

BigPandaToo commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/IndexPrivilegeIntegTests.java:
```diff
@@ -120,8 +121,9 @@ protected String configRoles() {
-        final String usersPasswdHashed = new String(Hasher.resolve(
-            randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9")).hash(new SecureString("passwd".toCharArray())));
+        final Hasher passwdHasher = inFipsJvm() ? Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000")) :
+            Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9"));
```
Same as above. "pbkdf2_stretch"?

jkakavas replied:
see previous comment

BigPandaToo commented on libs/ssl-config/src/test/java/org/elasticsearch/common/ssl/SslConfigurationLoaderTests.java:
```diff
@@ -103,6 +103,7 @@ public void testLoadTrustFromPemCAs() {
     }
 
     public void testLoadTrustFromPkcs12() {
+        assumeFalse("Can't use JKS/PKCS12 keystores in a FIPS JVM", inFipsJvm());
```
It may be outside of the scope of this PR, but do we plan to test Fips compatible PKCS12 keystores?

jkakavas replied:
There is no such thing as Fips compatible PKCS12 keystores, unfortunately*. See https://downloads.bouncycastle.org/fips-java/BC-FJA-UserGuide-1.0.2.pdf section 7 too. We document this in https://www.elastic.co/guide/en/elasticsearch/reference/current/fips-140-compliance.html#_tls_keystores_and_keys

\* Technically, I think it might be possible to create a PKCS12 implementation that uses only PBES2 ( i.e. PBKDF2 ) for key derivation, but that would be outside the scope of this PR and our FIPS 140 support too.

BigPandaToo replied:
Aha, I must have mistaken FIPS compatible with FIPS approved only compatible...

jkakavas replied:
Not sure what you mean by compatible here, we can also discuss live

jkakavas committed:
Fix password in rest tests

jkakavas committed:
adjust fips.gradle

jkakavas committed:
disable test that assumes an OSS distribution

jkakavas committed:
Mute tests that explicitly set the license to basic in FIPS 140, as fips mode is not supported in basic

jkakavas committed:
Fix SecurityDocumentationIT passwords

jkakavas committed:
Don't run ReloadSecureSettingsIT in fips mode. We could move this to xpack and enable TLS as a followup

jkakavas committed:
Don't run geo tests with default distribution testclusters

jkakavas committed:
Add password to the keystore of the docs testcluster, even when not in FIPS

jkakavas committed:
fix yaml syntax

jkakavas committed:
yet more unfixed passwords

jkakavas committed:
Merge remote-tracking branch 'origin/master' into fips-approved-mode-tests

jkakavas committed:
Fix SecurityIT

jkakavas committed:
correctly(this time) fix yaml syntax

jkakavas committed:
mute mapper-extras tests until issue with trial is resolved

jkakavas committed:
missing import

jkakavas committed:
Adjust passwords in docs integration tests

jkakavas committed:
Another round of password adjustments

jkakavas committed:
factor out selecting a random hashing algo to a method of its own

jkakavas committed:
revert unintended formatting changes

jkakavas committed:
Mute test that needs OSS cluster in FIPS 140 mode

BigPandaToo commented:
Also noticed x-pack/docs/build.gradle uses "test-password" in setup. Probably, need to change it too.

jkakavas commented:
> Also noticed x-pack/docs/build.gradle uses "test-password" in setup. Probably, need to change it too.

Can you point me to that @BigPandaToo ? I cant seem to find what you refer to

BigPandaToo reviewed (commented):
I left some commints/comments, overall looks good, we can discuss some details tomorrow.

jkakavas committed:
Add some information about testing in fips mode

jkakavas commented on client/rest-high-level/src/test/java/org/elasticsearch/client/SecurityIT.java:
```diff
@@ -222,7 +222,7 @@ private PutUserRequest randomPutUserRequest(boolean enabled) {
 
     private static PutUserRequest randomPutUserRequest(User user, boolean enabled) {
-        final char[] password = randomAlphaOfLengthBetween(6, 10).toCharArray();
+        final char[] password = randomAlphaOfLengthBetween(14, 19).toCharArray();
```
This is not only about verification. Put user requests will trigger password hashing, which will fail in fips mode with short passwords

BigPandaToo replied:
It doesn't look like it will. PutUserRequestBuilder does trigger a hasher, but we never get there. Looks like PutUserRequest is only used in tests.
Whith this said, I don't think it's a big problem to leave {14, 19} range here as long as we still have some tests covering the short allowed passwords when Fips is off.

jkakavas replied:
> It doesn't look like it will.

It will. 

Try `./gradlew client:rest-high-level:integTest --tests="org.elasticsearch.client.SecurityIT.testPutUser" in this branch which fails with passwords < 14 characters.

BigPandaToo replied:
Worked for me... Yes, I set the renge back to 6-10...

jkakavas replied:
This test suite sets up a test cluster (client/rest-high-level/build.gradle)  and `SecurityIT` performs security API related tests with the high level client against that cluster. `SecurityIT#testPutUser` creates a random PutUserRequest and _sends_ that request to the test cluster via the HLRC ( `SecurityClient securityClient = highLevelClient().security();` )
When the test cluster processes this request it will hash that user password in the process of storing it to the security index ( see `RestPutUserAction#innerPrepareRequest` which triggers the hashing ) . This hashing will fail when PBKDF2 is used and an exception will be thrown. 

If you don't get the failure, you either not run in FIPS mode ( `-Dtests.fips.enabled=true` ) or you don't run the test from this branch. You cannot try this in i.e. master as master doesn't yet contain [the fix](https://github.com/elastic/elasticsearch/pull/64024/files#diff-81f5504fc9e931a18720c9d0ecc27c0d9f610961e6e7c8e1fc072cfa0a11e5f0R56) which is the reason for this PR and sets the test cluster correctly in fips mode ) 

Hope this makes more sense now

jkakavas replied:
We agreed on why this change in the test is necessary over a chat, as such I'm resolving the comment

jkakavas commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SecurityDocumentationIT.java:
```diff
@@ -158,9 +158,9 @@ protected Settings restAdminSettings() {
+        addUser(client, usernames[0], randomAlphaOfLengthBetween(14, 18));
+        addUser(client, usernames[1], randomAlphaOfLengthBetween(14, 18));
+        addUser(client, usernames[2], randomAlphaOfLengthBetween(14, 18));
         {
```
Same as above, creating a user involves hashing the password for storage.

BigPandaToo replied:
Same as above...

jkakavas replied:
see above

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/user/PutUserRequestBuilderTests.java:
```diff
@@ -144,8 +145,8 @@ public void testWithEnabled() throws IOException {
     public void testWithValidPasswordHash() throws IOException {
-        final Hasher hasher = Hasher.BCRYPT4; // this is the fastest hasher we officially support
-        final char[] hash = hasher.hash(new SecureString("secret".toCharArray()));
+        final Hasher hasher = inFipsJvm() ? Hasher.PBKDF2_1000 : Hasher.BCRYPT4; // this is the fastest hasher we officially support
```
> Should we add some randomization here and below?

We don't aim to extensively test all hashing algorithms in this test so this is why the original only used BCRYPT4. There is no reason to not use getFastStoredHashAlgoForTests though, I'll change this

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/file/FileRealmTests.java:
```diff
@@ -61,21 +61,21 @@ public void init() throws Exception {
         userRolesStore = mock(FileUserRolesStore.class);
         globalSettings = Settings.builder().put("path.home", createTempDir()).put("xpack.security.authc.password_hashing.algorithm",
-            randomFrom("bcrypt9", "pbkdf2")).
+            inFipsJvm() ? "pbkdf2" : "bcrypt9").
```
same as above, I'll chnage this too

jkakavas committed:
address feedback

jkakavas committed:
Fix file user test

jkakavas committed:
Merge remote-tracking branch 'origin/master' into fips-approved-mode-tests

jkakavas commented:
@BigPandaToo this is ready for review

BigPandaToo commented:
> "test-password"

https://github.com/elastic/elasticsearch/blob/master/x-pack/docs/build.gradle#L708

jkakavas commented:
> > "test-password"
> 
> https://github.com/elastic/elasticsearch/blob/master/x-pack/docs/build.gradle#L708

I think you are looking at a different branch, this is already changed in this PR: https://github.com/elastic/elasticsearch/pull/64024/files#diff-c6a26b2baddac45416af5df48277316a61f3fede5092daf25300ff36b7cd4f00R708

BigPandaToo reviewed (approved):
I left some comments regarding the passwords in tests. I still don't think we need to change all of them, but with that being said I don't think it is important, so I leave it for you to decide.
LGTM otherwise

jkakavas commented:
@elasticmachine run elasticsearch-ci/2

jkakavas commented:
@ywangd can you please take a look :pray:

ywangd reviewed (approved):
LGTM

@jkakavas Thanks for working on this. It is an admirable effort! Even browsing through it takes some time. I left a few comments, nothing really major. Also:
* I wonder whether we should ask or work with infra to have a new elasticmachine command for running FIPS tests against a PR. I don't know how you managed it (did you run it locally?), but I would have struggled without it.
* Just a friendly reminder, in case it is not tracked anywhere, we could use some tests for #65464 after this PR is merged.
* It is relatively easy to violate the password length rule in future new tests if the author is not aware of FIPS requirement (which is likely). And because normal PR CI does not run FIPS, there is no way for him/her to know about it untill it fails in the periodical FIPS CI tests. It will then likely be our responsibillity to play whack a mole. Therefore, I wonder where there could be a way to enforce/detect it during regular test time. Something like having a test only system property and ask hasher to fail if it sees the property and a short password. Anyway, I think it should be a separate PR if we decide it is something we want to have.

jkakavas commented on modules/mapper-extras/build.gradle:
```diff
@@ -29,3 +31,8 @@ restResources {
   }
 }
+if (BuildParams.inFipsJvm){
+  // The cluster fails to form with a trial license and we can't test FIPS 140 mode with a basic license
```
The description was vague because I hadn't figured why either. I reached out to core-features and this is now resolved in https://github.com/elastic/elasticsearch/pull/65971, so I will amend this

jkakavas commented on modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/71_context_api.yml:
```diff
@@ -1,11 +1,17 @@
 "Action to list contexts":
+    - skip:
+        features: fips_140
+        reason: "The tests expect to be run with OSS distribution"
```
The problem for these tests is that these expect to run in a cluster with the OSS distribution and the expected number and order of contexts, and classes order and number for the score context are different between oss and default distribution test clusters. In FIPS 140 mode we use the default distribution ( because i.e. one can't set `xpack.security.fips_mode.enabled:` in OSS ) , so these tests would fail.

jkakavas commented on plugins/examples/painless-whitelist/build.gradle:
```diff
@@ -31,7 +33,10 @@ esplugin {
   compileOnly "org.elasticsearch.plugin:elasticsearch-scripting-painless-spi:${versions.elasticsearch}"
 }
-
+if (BuildParams.inFipsJvm){
+  // This test cluster is using OSS distribution and FIPS 140 mode is not supported in OSS
+  tasks.named("yamlRestTest").configure{enabled = false }
+}
```
Good point, this must be a left over from a previous commit

jkakavas commented on server/src/internalClusterTest/java/org/elasticsearch/action/admin/ReloadSecureSettingsIT.java:
```diff
@@ -54,6 +55,12 @@
 public class ReloadSecureSettingsIT extends ESIntegTestCase {
 
+    @BeforeClass
+    public static void disableInFips() {
+        assumeFalse("Cannot run in FIPS mode since the keystore will be password protected and sending a password in the reload" +
+            "settings api call, require TLS to be configured for the transport layer", inFipsJvm());
+    }
```
will do , good idea

jkakavas commented on x-pack/plugin/enrich/qa/rest-with-security/build.gradle:
```diff
@@ -1,9 +1,15 @@
   javaRestTestImplementation project(path: xpackModule('core'))
   javaRestTestImplementation project(path: xpackModule('enrich:qa:common'))
 }
+if (BuildParams.inFipsJvm){
+  // This test cluster is using a BASIC license and FIPS 140 mode is not supported in BASIC
+  tasks.named("javaRestTest").configure{enabled = false }
+}
```
> Does our FIPS mode cover the plugins as well?

It does. Our plugins that is. For instance, we document that ingest-attachment is not supported. 

Let's handle this in a follow up. It wasn't obvious for QA tests that use basic license if this is on purpose for some specific reason, copy paste from other QA setups, or an arbitrary decision and reaching out to folks to figure this out is time-consuming. I elected to mute these for now, and we can revisit in a follow up.

jkakavas commented:
>  I wonder whether we should ask or work with infra to have a new elasticmachine command for running FIPS tests against a PR. I don't know how you managed it (did you run it locally?), but I would have struggled without it.

Yes, it's in my TODO list to do something about this, I was discussing this with @rdnm a couple of days ago actually. So far, it has been only my raising fips related PRs and I happen to have a workstation where I run CI on but this is not a scalable solution. 

> Just a friendly reminder, in case it is not tracked anywhere, we could use some tests for #65464 after this PR is merged.

It's not tracked in an issue, @BigPandaToo mentioned it to me yesterday too. I plan on raising a small PR soon, that's why I haven't raised an issue. 

>  It is relatively easy to violate the password length rule in future new tests if the author is not aware of FIPS requirement (which is likely). And because normal PR CI does not run FIPS, there is no way for him/her to know about it untill it fails in the periodical FIPS CI tests. It will then likely be our responsibillity to play whack a mole. 

yeap agreed

> > The approach
> > is brittle as there is no guarantee that the next test that will be added won't
> > use a short password, so we add some testing documentation too.

> Therefore, I wonder where there could be a way to enforce/detect it during regular test time. Something like having a test only system property and ask hasher to fail if it sees the property and a short password. Anyway, I think it should be a separate PR if we decide it is something we want to have.

I don't like the idea of having CI related code in our production code but we've been whack-a-moling for a long time with fips CI . A solution is probably a mix of developer enablement and CI checks, and we should figure the right balance - but in a followup, as you say

jkakavas committed:
Merge remote-tracking branch 'origin/master' into fips-approved-mode-tests

ywangd commented:
> I don't like the idea of having CI related code in our production code but we've been whack-a-moling for a long time with fips CI . 
>
I was thinking about something like
```java
assert Boolean.parseBoolean(System.getProperty("is.in.a.test", "false")) == false || data.length() >= 14
            : "Passwords for tests must be longer than 14 char";
```
in a method like `Hasher#getPbkdf2Hash`. Or maybe just
```java
assert data.length() >= 14
```
is enough because `assert` only runs in tests. We have lots of other asserts in the code base. So it feels reasonable to me.

jkakavas commented:
We'd need to guard this with an `inFipsJvm` though ( which would be reading a sysprop that we use only in CI), because someone might want to write a test that uses a shorter password on purpose and mute it in fips mode

jkakavas committed:
address feedback

ywangd commented:
> We'd need to guard this with an `inFipsJvm` though ( which would be reading a sysprop that we use only in CI), because someone might want to write a test that uses a shorter password on purpose and mute it in fips mode

Guarding it inside `inFipsJvm` would make the assertion pointless because it meant to catch issue without having to run FIPS (as is the case for PR CI). But your point of "a test that uses a shorter password on purpose and mute it in fips mode" is a possible scenario, though I cannot think of why people would absolutely need a short password. Anyway, in this case, it is not possible to have an automated solution.

jkakavas committed:
Expect warnings when running with default distribution

jkakavas committed:
Use strong passwords in tests

jkakavas committed:
Use appropriate length private keys

jkakavas committed:
Use appropriate length SecretKey for EC2, as it is used for key generation and should be > 112 bits in fips mode

jkakavas committed:
Merge remote-tracking branch 'origin/master' into fips-approved-mode-tests

jkakavas committed:
Change GCS RSA keys to 2048bits so that tests don't fail in FIPS mode

jkakavas committed:
Mute CCR test downgrading to basic in FIPS mode

jkakavas committed:
fix more

jkakavas committed:
Mute tests in fips when security is disabled

jkakavas committed:
Mute test using PKCS12

jkakavas committed:
remove unnecessary policy application

jkakavas committed:
Merge remote-tracking branch 'origin/master' into fips-approved-mode-tests

jkakavas committed:
Ensure Kerberos fixture runs with fips 140-2 approved enctypes

jkakavas committed:
Allow ESRestTestClient to trust certificates

ESRestTestCase rest clients could only be configured to trust
the certificate authorities that were contained in a truststore. In
certain cases (like in fips mode where JKS/PKCS12 keystores) cannot
be used, it's beneficial to be able to trust specific certificate
authorities (indicated by the CA PEM endoded certificate)

jkakavas committed:
use long passwords in kerberos tests

jkakavas committed:
adjust password length

jkakavas committed:
mute tests instead of all the tasks

jkakavas committed:
Change moar passwords

jkakavas committed:
Mutes qa tests that fail in CI until we can investigate more

jkakavas committed:
yet another password

jkakavas committed:
no tests run

jkakavas committed:
Adjust reindex from remote QA test to use PEM CA cert instead of a PKCS12 store

jkakavas committed:
more passwords

jkakavas committed:
reference issue in muted CCR tests

jkakavas committed:
Merge remote-tracking branch 'origin/master' into fips-approved-mode-tests

jkakavas committed:
handle encrypted keystores in SetupPasswordToolIT

jkakavas committed:
set license to trial for testclusters as we can't use fips mode in basic. All the tests that require basic license are already muted in fips mode so they won't be affected

jkakavas committed:
Disable bwc tests in fips mode as there arent' many ( only 7.11 ) versions that can be run in a correctly configured FIPS 140-2 JVM and we are lacking granural control of which bwc versions can be used

jkakavas committed:
use long enough azure.client.key in tests

jkakavas committed:
use long enough ec2 access keys in tests

jkakavas committed:
another short access key

jkakavas committed:
Merge remote-tracking branch 'origin/master' into fips-approved-mode-tests

jkakavas committed:
Change docker integ tests to use PEM key/certificates instead of keystores so that these can be run in fips mode

jkakavas committed:
Change remote-clusters integ tests to use PEM key/certificates instead of keystores so that these can be run in fips mode

jkakavas committed:
Fix all s3 fixture related tests by using the correct access key

jkakavas committed:
Merge remote-tracking branch 'origin/master' into fips-approved-mode-tests
